src/crawlee/storages/_request_source_tandem.py

```diff
@@ -0,0 +1,109 @@
+from __future__ import annotations
+
+from datetime import timedelta
+from logging import getLogger
+from typing import TYPE_CHECKING
+
+from typing_extensions import override
+
+from crawlee.storages._request_provider import RequestProvider
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+
+    from crawlee import Request
+    from crawlee.base_storage_client._models import ProcessedRequest
+    from crawlee.storages._request_source import RequestSource
+
+
+logger = getLogger(__name__)
+
+
+class RequestSourceTandem(RequestProvider):
+    """Implements a tandem behaviour for a pair of `RequestSource` and `RequestProvider`.
+
+    In this scenario, the contents of the "source" get transferred into the "provider", allowing processing the requests
+    from both sources and also enqueueing new requests (not possible with plain `RequestSource`).
+    """
+
+    def __init__(self, request_source: RequestSource, request_provider: RequestProvider) -> None:
+        self._read_only_source = request_source
+        self._read_write_provider = request_provider
```

Comment:
I don't know, but the fact that we both inherit from `RequestProvider` and also accept it as an input argument suggests there might be a design issue. Isn't it?

Reply:
Well, what the `RequestSourceTandem` does is that it takes a `RequestProvider`, duct-tapes a `RequestSource` on top of it and then acts as a `RequestProvider` itself. I'm sure it's a known design pattern with a name and everything :smile:

Reply:
Yeah, the source of my concerns probably arise from the fact, that Python uses the same syntax for "implement an interface" and regular inheritance.